File: src/shared/application/services/positioning/arrows/placement/default_placement_service.py
"""
Default Placement Service

This service loads and uses default placement JSON files for accurate
arrow positioning with comprehensive adjustment data.
"""

import codecs
import json
from pathlib import Path
from typing import Any, Optional

# Use framework-agnostic types from desktop.modern.core.types
from desktop.modern.core.types import Point
from desktop.modern.domain.models import MotionData, MotionType


class DefaultPlacementService:
    """Service that loads default placement data and provides adjustments."""

    def __init__(self):
        self.root_path = self._find_project_root()
        self.all_defaults: dict[str, dict[str, dict[str, Any]]] = {
            "diamond": {},
            "box": {},
        }

        self.placements_files = {
            "diamond": {
                "pro": "default/default_diamond_pro_placements.json",
                "anti": "default/default_diamond_anti_placements.json",
                "float": "default/default_diamond_float_placements.json",
                "dash": "default/default_diamond_dash_placements.json",
                "static": "default/default_diamond_static_placements.json",
            },
            "box": {
                "pro": "default/default_box_pro_placements.json",
                "anti": "default/default_box_anti_placements.json",
                "float": "default/default_box_float_placements.json",
                "dash": "default/default_box_dash_placements.json",
                "static": "default/default_box_static_placements.json",
            },
        }

        self._load_all_default_placements()

    def _find_project_root(self) -> Path:
        """Automatically find the project root by looking for pyproject.toml or .git."""
        current = Path(__file__).resolve()
        for parent in [current] + list(current.parents):
            if (parent / "pyproject.toml").exists() or (parent / ".git").exists():
                return parent
        return current.parent  # fallback

    def _load_all_default_placements(self) -> None:
        """Load all default placement JSON files using direct path resolution."""
        for grid_mode, motion_files in self.placements_files.items():
            for motion_type, filename in motion_files.items():
                filepath = (
                    self.root_path / "data" / "arrow_placement" / grid_mode / filename
                )
                self.all_defaults[grid_mode][motion_type] = self._load_json(
                    str(filepath)
                )

    def _load_json(self, path: str) -> dict[str, Any]:
        """Load JSON file with error handling."""
        try:
            with codecs.open(path, "r", encoding="utf-8") as file:
                return json.load(file)
        except Exception as e:
            print(f"Error loading default placements from {path}: {e}")
            return {}

    def get_default_adjustment(
        self,
        motion_data: MotionData,
        grid_mode: str = "diamond",
        placement_key: Optional[str] = None,
    ) -> Point:
        """
        Get default adjustment using validated placement system.

        Args:
            motion_data: The motion data
            grid_mode: Grid mode (diamond or box)
            placement_key: Specific placement key (e.g., "pro_to_layer1_alpha")
                          If None, will use basic motion type

        Returns:
            Point with adjustment values
        """
        if grid_mode not in ["diamond", "box"]:
            grid_mode = "diamond"

        motion_type_str = motion_data.motion_type.value
        default_placements = self.all_defaults.get(grid_mode, {}).get(
            motion_type_str, {}
        )

        if not default_placements:
            return Point(0, 0)

        # If specific placement key provided, use it
        if placement_key and placement_key in default_placements:
            adjustment_key = placement_key
        else:
            # Use basic motion type as fallback
            adjustment_key = motion_type_str

        # Get the adjustment values for the specific turns
        # Convert turns to the format used in JSON files (no decimal for whole numbers)
        turns_value = motion_data.turns
        if turns_value == int(turns_value):
            turns_str = str(int(turns_value))  # "1.0" → "1"
        else:
            turns_str = str(turns_value)  # "0.5" → "0.5"

        adjustment_data = default_placements.get(adjustment_key, {})

        if isinstance(adjustment_data, dict):
            x, y = adjustment_data.get(turns_str, (0, 0))
        else:
            x, y = (0, 0)

        return Point(x, y)

    def get_available_placement_keys(
        self, motion_type: MotionType, grid_mode: str = "diamond"
    ) -> list[str]:
        """Get all available placement keys for a motion type."""
        motion_type_str = motion_type.value
        default_placements = self.all_defaults.get(grid_mode, {}).get(
            motion_type_str, {}
        )
        return list(default_placements.keys())

    def get_placement_data(
        self, motion_type: MotionType, placement_key: str, grid_mode: str = "diamond"
    ) -> dict[str, tuple[int, int]]:
        """Get complete placement data for a specific key."""
        motion_type_str = motion_type.value
        default_placements = self.all_defaults.get(grid_mode, {}).get(
            motion_type_str, {}
        )
        return default_placements.get(placement_key, {})

    def debug_available_keys(
        self, motion_type: MotionType, grid_mode: str = "diamond"
    ) -> None:
        """Debug helper to print all available placement keys."""
        keys = self.get_available_placement_keys(motion_type, grid_mode)
        print(f"Available placement keys for {motion_type.value} in {grid_mode} mode:")
        for key in keys:
            placement_data = self.get_placement_data(motion_type, key, grid_mode)
            print(f"  {key}: {placement_data}")

File: src/web/web_app/src/lib/services/positioning/arrows/placement/DefaultPlacementService.ts
/**
 * Default Placement Service
 *
 * Provides default arrow placement adjustments by loading data from JSON files.
 * Mirrors the exact functionality from desktop DefaultPlacementService.
 */

import { ArrowPlacementDataService } from '../../../implementations/ArrowPlacementDataService';
import type { GridMode, MotionType } from '../../../interfaces';

/**
 * Interface for Default Placement Service that mirrors Python implementation
 */
export interface IDefaultPlacementServiceJson {
	getDefaultAdjustment(
		placementKey: string,
		turns: number | string,
		motionType: MotionType,
		gridMode: GridMode
	): Promise<{ x: number; y: number }>;

	getAvailablePlacementKeys(motionType: MotionType, gridMode: GridMode): Promise<string[]>;

	isLoaded(): boolean;

	getPlacementData(
		motionType: MotionType,
		placementKey: string,
		gridMode: GridMode
	): Promise<{ [turns: string]: [number, number] }>;

	debugAvailableKeys(motionType: MotionType, gridMode: GridMode): Promise<void>;
}

export class DefaultPlacementService implements IDefaultPlacementServiceJson {
	private placementDataService: ArrowPlacementDataService;

	constructor() {
		this.placementDataService = new ArrowPlacementDataService();
	}

	/**
	 * Get default adjustment for arrow placement using placement key and turns.
	 * This mirrors the Python get_default_adjustment() method.
	 *
	 * @param placementKey - The placement identifier
	 * @param turns - Number of turns or turn identifier
	 * @param motionType - Motion type (pro, anti, float, dash, static)
	 * @param gridMode - Grid mode (diamond, box)
	 * @returns Adjustment coordinates {x, y}
	 */
	async getDefaultAdjustment(
		placementKey: string,
		turns: number | string,
		motionType: MotionType,
		gridMode: GridMode
	): Promise<{ x: number; y: number }> {
		console.log(
			`DefaultPlacementService.getDefaultAdjustment() called with:`,
			`placement_key=${placementKey}, turns=${turns}, motion_type=${motionType}, grid_mode=${gridMode}`
		);

		try {
			// Ensure placement data is loaded
			await this._loadAllDefaultPlacements();

			// Get the adjustment from the data service
			const adjustment = await this.placementDataService.getDefaultAdjustment(
				motionType,
				placementKey,
				turns,
				gridMode
			);

			console.log(
				`Found default adjustment for ${placementKey} at ${turns} turns: [${adjustment.x}, ${adjustment.y}]`
			);

			return adjustment;
		} catch (error) {
			console.warn(
				`Failed to get default adjustment for ${placementKey} at ${turns} turns:`,
				error
			);
			return { x: 0, y: 0 };
		}
	}

	/**
	 * Get available placement keys for a given motion type and grid mode.
	 * This mirrors the Python get_available_placement_keys() method.
	 *
	 * @param motionType - Motion type (pro, anti, float, dash, static)
	 * @param gridMode - Grid mode (diamond, box)
	 * @returns Array of available placement key strings
	 */
	async getAvailablePlacementKeys(motionType: MotionType, gridMode: GridMode): Promise<string[]> {
		await this._loadAllDefaultPlacements();
		return this.placementDataService.getAvailablePlacementKeys(motionType, gridMode);
	}

	/**
	 * Check if placement data has been loaded.
	 * This mirrors the Python is_loaded() method.
	 *
	 * @returns true if data is loaded, false otherwise
	 */
	isLoaded(): boolean {
		return this.placementDataService.isLoaded();
	}

	/**
	 * Load all default placement data from JSON files.
	 * This mirrors the Python _load_all_default_placements() method.
	 *
	 * The Python version loads from:
	 * - /data/arrow_placement/{grid_mode}/default/{motion_type}_placements.json
	 *
	 * Our TypeScript version uses the same file structure and loading pattern.
	 */
	private async _loadAllDefaultPlacements(): Promise<void> {
		if (this.placementDataService.isLoaded()) {
			return;
		}

		console.log('DefaultPlacementService: Loading all default placements...');

		try {
			await this.placementDataService.loadPlacementData();
			console.log('✅ DefaultPlacementService: All placement data loaded successfully');
		} catch (error) {
			console.error('❌ DefaultPlacementService: Failed to load placement data:', error);
			throw new Error(
				`Default placement loading failed: ${error instanceof Error ? error.message : 'Unknown error'}`
			);
		}
	}

	/**
	 * Get raw placement data for debugging purposes.
	 * This mirrors the Python get_placement_data() method.
	 *
	 * @param motionType - Motion type (pro, anti, float, dash, static)
	 * @param placementKey - The placement identifier
	 * @param gridMode - Grid mode (diamond, box)
	 * @returns Raw placement data mapping turns to [x, y] adjustments
	 */
	async getPlacementData(
		motionType: MotionType,
		placementKey: string,
		gridMode: GridMode
	): Promise<{ [turns: string]: [number, number] }> {
		await this._loadAllDefaultPlacements();
		return this.placementDataService.getPlacementData(motionType, placementKey, gridMode);
	}

	/**
	 * Debug method to log available placement keys.
	 * This mirrors the Python debug_available_keys() method.
	 *
	 * @param motionType - Motion type to debug
	 * @param gridMode - Grid mode to debug
	 */
	async debugAvailableKeys(motionType: MotionType, gridMode: GridMode): Promise<void> {
		await this._loadAllDefaultPlacements();
		await this.placementDataService.debugAvailableKeys(motionType, gridMode);
	}
}
